// @datadog/browser-rum/cjs/domain/record/internalApi.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.takeFullSnapshot = takeFullSnapshot;
exports.takeNodeSnapshot = takeNodeSnapshot;
const browser_core_1 = require("@datadog/browser-core");
const browser_rum_core_1 = require("@datadog/browser-rum-core");
const startFullSnapshots_1 = require("./startFullSnapshots");
const recordingScope_1 = require("./recordingScope");
const elementsScrollPositions_1 = require("./elementsScrollPositions");
const serialization_1 = require("./serialization");
/**
 * Take a full snapshot of the document, generating the same records that the browser SDK
 * would generate.
 *
 * This is an internal API function. Be sure to update Datadog-internal callers if you
 * change its signature or behavior.
 */
function takeFullSnapshot({ configuration, } = {}) {
    const records = [];
    const emitRecord = (record) => {
        records.push(record);
    };
    (0, startFullSnapshots_1.takeFullSnapshot)((0, browser_core_1.timeStampNow)(), 0 /* SerializationKind.INITIAL_FULL_SNAPSHOT */, emitRecord, browser_core_1.noop, createTemporaryRecordingScope(configuration));
    return records;
}
/**
 * Take a snapshot of a DOM node, generating the serialized representation that the
 * browser SDK would generate.
 *
 * This is an internal API function. Be sure to update Datadog-internal callers if you
 * change its signature or behavior.
 */
function takeNodeSnapshot(node, { configuration } = {}) {
    let serializedNode = null;
    (0, serialization_1.serializeInTransaction)(0 /* SerializationKind.INITIAL_FULL_SNAPSHOT */, browser_core_1.noop, browser_core_1.noop, createTemporaryRecordingScope(configuration), (transaction) => {
        const privacyLevel = (0, browser_rum_core_1.getNodePrivacyLevel)(node, transaction.scope.configuration.defaultPrivacyLevel);
        if (privacyLevel === browser_rum_core_1.NodePrivacyLevel.HIDDEN || privacyLevel === browser_rum_core_1.NodePrivacyLevel.IGNORE) {
            return;
        }
        serializedNode = (0, serialization_1.serializeNode)(node, privacyLevel, transaction);
    });
    return serializedNode;
}
function createTemporaryRecordingScope(configuration) {
    return (0, recordingScope_1.createRecordingScope)({
        defaultPrivacyLevel: browser_rum_core_1.NodePrivacyLevel.ALLOW,
        ...configuration,
    }, (0, elementsScrollPositions_1.createElementsScrollPositions)(), {
        addShadowRoot: browser_core_1.noop,
        removeShadowRoot: browser_core_1.noop,
        flush: browser_core_1.noop,
        stop: browser_core_1.noop,
    });
}
