// @datadog/browser-rum/cjs/domain/record/itemIds.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createEventIds = createEventIds;
exports.createNodeIds = createNodeIds;
exports.createStringIds = createStringIds;
exports.createStyleSheetIds = createStyleSheetIds;
function createEventIds() {
    return createWeakIdMap(1 /* EventIdConstants.FIRST_ID */);
}
function createNodeIds() {
    return createWeakIdMap(0 /* NodeIdConstants.FIRST_ID */);
}
function createStringIds() {
    return createIdMap(0 /* StringIdConstants.FIRST_ID */);
}
function createStyleSheetIds() {
    return createWeakIdMap(0 /* StyleSheetIdConstants.FIRST_ID */);
}
function createIdMap(firstId) {
    return createItemIds(() => new Map(), firstId);
}
function createWeakIdMap(firstId) {
    return createItemIds(() => new WeakMap(), firstId);
}
function createItemIds(createMap, firstId) {
    let map = createMap();
    let nextId = firstId;
    const get = (object) => map.get(object);
    return {
        clear() {
            if (nextId === firstId) {
                return;
            }
            map = createMap();
            nextId = firstId;
        },
        delete(object) {
            map.delete(object);
        },
        get,
        getOrInsert(object) {
            // Try to reuse any existing id.
            let id = get(object);
            if (id === undefined) {
                id = nextId++;
                map.set(object, id);
            }
            return id;
        },
        get nextId() {
            return nextId;
        },
        get size() {
            return nextId - firstId;
        },
    };
}
