// @datadog/browser-rum/cjs/domain/record/startFullSnapshots.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.startFullSnapshots = startFullSnapshots;
exports.takeFullSnapshot = takeFullSnapshot;
const browser_rum_core_1 = require("@datadog/browser-rum-core");
const browser_core_1 = require("@datadog/browser-core");
const types_1 = require("../../types");
const serialization_1 = require("./serialization");
const viewports_1 = require("./viewports");
function startFullSnapshots(lifeCycle, emitRecord, emitStats, flushMutations, scope, serialize = defaultSerializeFullSnapshotCallback()) {
    takeFullSnapshot((0, browser_core_1.timeStampNow)(), 0 /* SerializationKind.INITIAL_FULL_SNAPSHOT */, emitRecord, emitStats, scope, serialize);
    const { unsubscribe } = lifeCycle.subscribe(2 /* LifeCycleEventType.VIEW_CREATED */, (view) => {
        flushMutations();
        takeFullSnapshot(view.startClocks.timeStamp, 1 /* SerializationKind.SUBSEQUENT_FULL_SNAPSHOT */, emitRecord, emitStats, scope, serialize);
    });
    return {
        stop: unsubscribe,
    };
}
function takeFullSnapshot(timestamp, kind, emitRecord, emitStats, scope, serialize = defaultSerializeFullSnapshotCallback()) {
    const { width, height } = (0, browser_rum_core_1.getViewportDimension)();
    emitRecord({
        data: {
            height,
            href: window.location.href,
            width,
        },
        type: types_1.RecordType.Meta,
        timestamp,
    });
    emitRecord({
        data: {
            has_focus: document.hasFocus(),
        },
        type: types_1.RecordType.Focus,
        timestamp,
    });
    serialize(timestamp, kind, document, emitRecord, emitStats, scope);
    if (window.visualViewport) {
        emitRecord({
            data: (0, viewports_1.getVisualViewport)(window.visualViewport),
            type: types_1.RecordType.VisualViewport,
            timestamp,
        });
    }
}
function defaultSerializeFullSnapshotCallback() {
    return (0, browser_core_1.isExperimentalFeatureEnabled)(browser_core_1.ExperimentalFeature.USE_CHANGE_RECORDS)
        ? serialization_1.serializeFullSnapshotAsChange
        : serialization_1.serializeFullSnapshot;
}
