// @datadog/browser-rum-core/cjs/domain/action/trackManualActions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.trackManualActions = trackManualActions;
const browser_core_1 = require("@datadog/browser-core");
const rawRumEvent_types_1 = require("../../rawRumEvent.types");
const eventTracker_1 = require("../eventTracker");
const isActionChildEvent_1 = require("./isActionChildEvent");
function trackManualActions(lifeCycle, onManualActionCompleted) {
    const actionTracker = (0, eventTracker_1.startEventTracker)(lifeCycle);
    function startManualAction(name, options = {}, startClocks = (0, browser_core_1.clocksNow)()) {
        var _a;
        const lookupKey = (_a = options.actionKey) !== null && _a !== void 0 ? _a : name;
        actionTracker.start(lookupKey, startClocks, {
            name,
            ...options,
        }, { isChildEvent: isActionChildEvent_1.isActionChildEvent });
    }
    function stopManualAction(name, options = {}, stopClocks = (0, browser_core_1.clocksNow)()) {
        var _a;
        const lookupKey = (_a = options.actionKey) !== null && _a !== void 0 ? _a : name;
        const stopped = actionTracker.stop(lookupKey, stopClocks, options);
        if (!stopped) {
            return;
        }
        const frustrationTypes = [];
        if (stopped.counts && stopped.counts.errorCount > 0) {
            frustrationTypes.push(rawRumEvent_types_1.FrustrationType.ERROR_CLICK);
        }
        const manualAction = {
            ...stopped,
            type: stopped.type || rawRumEvent_types_1.ActionType.CUSTOM,
            frustrationTypes,
        };
        onManualActionCompleted(manualAction);
    }
    function addInstantAction(action) {
        onManualActionCompleted({ id: (0, browser_core_1.generateUUID)(), frustrationTypes: [], ...action });
    }
    return {
        addAction: addInstantAction,
        startAction: startManualAction,
        stopAction: stopManualAction,
        findActionId: actionTracker.findId,
        stop: actionTracker.stopAll,
    };
}
