// @datadog/browser-rum-core/cjs/domain/action/trackClickActions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.trackClickActions = trackClickActions;
exports.finalizeClicks = finalizeClicks;
const browser_core_1 = require("@datadog/browser-core");
const htmlDomUtils_1 = require("../../browser/htmlDomUtils");
const rawRumEvent_types_1 = require("../../rawRumEvent.types");
const waitPageActivityEnd_1 = require("../waitPageActivityEnd");
const getSelectorFromElement_1 = require("../getSelectorFromElement");
const privacy_1 = require("../privacy");
const privacyConstants_1 = require("../privacyConstants");
const eventTracker_1 = require("../eventTracker");
const clickChain_1 = require("./clickChain");
const getActionNameFromElement_1 = require("./getActionNameFromElement");
const listenActionEvents_1 = require("./listenActionEvents");
const computeFrustration_1 = require("./computeFrustration");
const interactionSelectorCache_1 = require("./interactionSelectorCache");
const isActionChildEvent_1 = require("./isActionChildEvent");
function trackClickActions(lifeCycle, domMutationObservable, windowOpenObservable, configuration) {
    const actionTracker = (0, eventTracker_1.startEventTracker)(lifeCycle);
    const stopObservable = new browser_core_1.Observable();
    let currentClickChain;
    lifeCycle.subscribe(5 /* LifeCycleEventType.VIEW_ENDED */, stopClickChain);
    lifeCycle.subscribe(11 /* LifeCycleEventType.PAGE_MAY_EXIT */, stopClickChain);
    const { stop: stopActionEventsListener } = (0, listenActionEvents_1.listenActionEvents)(configuration, {
        onPointerDown: (pointerDownEvent) => processPointerDown(configuration, lifeCycle, domMutationObservable, pointerDownEvent, windowOpenObservable),
        onPointerUp: ({ clickActionBase, hadActivityOnPointerDown }, startEvent, getUserActivity) => {
            startClickAction(configuration, lifeCycle, domMutationObservable, windowOpenObservable, actionTracker, stopObservable, appendClickToClickChain, clickActionBase, startEvent, getUserActivity, hadActivityOnPointerDown);
        },
    });
    return {
        stop: () => {
            stopClickChain();
            stopObservable.notify();
            stopActionEventsListener();
            actionTracker.stopAll();
        },
        findActionId: actionTracker.findId,
    };
    function appendClickToClickChain(click) {
        if (!currentClickChain || !currentClickChain.tryAppend(click)) {
            const rageClick = click.clone();
            currentClickChain = (0, clickChain_1.createClickChain)(click, (clicks) => {
                finalizeClicks(clicks, rageClick);
                // Clear the reference to allow garbage collection. Without this, the finalize callback
                // retains a closure reference to the old click chain, preventing it from being cleaned up
                // and causing a memory leak as click chains accumulate over time.
                currentClickChain = undefined;
            });
        }
    }
    function stopClickChain() {
        if (currentClickChain) {
            currentClickChain.stop();
        }
    }
}
function processPointerDown(configuration, lifeCycle, domMutationObservable, pointerDownEvent, windowOpenObservable) {
    const targetForPrivacy = configuration.betaTrackActionsInShadowDom
        ? getEventTarget(pointerDownEvent)
        : pointerDownEvent.target;
    let nodePrivacyLevel;
    if (configuration.enablePrivacyForActionName) {
        nodePrivacyLevel = (0, privacy_1.getNodePrivacyLevel)(targetForPrivacy, configuration.defaultPrivacyLevel);
    }
    else {
        nodePrivacyLevel = privacyConstants_1.NodePrivacyLevel.ALLOW;
    }
    if (nodePrivacyLevel === privacyConstants_1.NodePrivacyLevel.HIDDEN) {
        return undefined;
    }
    const clickActionBase = computeClickActionBase(pointerDownEvent, nodePrivacyLevel, configuration);
    let hadActivityOnPointerDown = false;
    (0, waitPageActivityEnd_1.waitPageActivityEnd)(lifeCycle, domMutationObservable, windowOpenObservable, configuration, (pageActivityEndEvent) => {
        hadActivityOnPointerDown = pageActivityEndEvent.hadActivity;
    }, 
    // We don't care about the activity duration, we just want to know whether an activity did happen
    // within the "validation delay" or not. Limit the duration so the callback is called sooner.
    waitPageActivityEnd_1.PAGE_ACTIVITY_VALIDATION_DELAY);
    return { clickActionBase, hadActivityOnPointerDown: () => hadActivityOnPointerDown };
}
function startClickAction(configuration, lifeCycle, domMutationObservable, windowOpenObservable, actionTracker, stopObservable, appendClickToClickChain, clickActionBase, startEvent, getUserActivity, hadActivityOnPointerDown) {
    var _a;
    const click = newClick(lifeCycle, actionTracker, getUserActivity, clickActionBase, startEvent);
    appendClickToClickChain(click);
    const selector = (_a = clickActionBase === null || clickActionBase === void 0 ? void 0 : clickActionBase.target) === null || _a === void 0 ? void 0 : _a.selector;
    if (selector) {
        (0, interactionSelectorCache_1.updateInteractionSelector)(startEvent.timeStamp, selector);
    }
    const { stop: stopWaitPageActivityEnd } = (0, waitPageActivityEnd_1.waitPageActivityEnd)(lifeCycle, domMutationObservable, windowOpenObservable, configuration, (pageActivityEndEvent) => {
        if (pageActivityEndEvent.hadActivity && pageActivityEndEvent.end < click.startClocks.timeStamp) {
            // If the clock is looking weird, just discard the click
            click.discard();
        }
        else {
            if (pageActivityEndEvent.hadActivity) {
                click.stop(pageActivityEndEvent.end);
            }
            else if (hadActivityOnPointerDown()) {
                click.stop(
                // using the click start as activity end, so the click will have some activity but its
                // duration will be 0 (as the activity started before the click start)
                click.startClocks.timeStamp);
            }
            else {
                click.stop();
            }
        }
    }, interactionSelectorCache_1.CLICK_ACTION_MAX_DURATION);
    const viewEndedSubscription = lifeCycle.subscribe(5 /* LifeCycleEventType.VIEW_ENDED */, ({ endClocks }) => {
        click.stop(endClocks.timeStamp);
    });
    const pageMayExitSubscription = lifeCycle.subscribe(11 /* LifeCycleEventType.PAGE_MAY_EXIT */, () => {
        click.stop((0, browser_core_1.timeStampNow)());
    });
    const stopSubscription = stopObservable.subscribe(() => {
        click.stop();
    });
    click.stopObservable.subscribe(() => {
        pageMayExitSubscription.unsubscribe();
        viewEndedSubscription.unsubscribe();
        stopWaitPageActivityEnd();
        stopSubscription.unsubscribe();
    });
}
function computeClickActionBase(event, nodePrivacyLevel, configuration) {
    const target = configuration.betaTrackActionsInShadowDom ? getEventTarget(event) : event.target;
    const rect = target.getBoundingClientRect();
    const selector = (0, getSelectorFromElement_1.getSelectorFromElement)(target, configuration.actionNameAttribute);
    if (selector) {
        (0, interactionSelectorCache_1.updateInteractionSelector)(event.timeStamp, selector);
    }
    const { name, nameSource } = (0, getActionNameFromElement_1.getActionNameFromElement)(target, configuration, nodePrivacyLevel);
    return {
        type: rawRumEvent_types_1.ActionType.CLICK,
        target: {
            width: Math.round(rect.width),
            height: Math.round(rect.height),
            selector,
        },
        position: {
            // Use clientX and Y because for SVG element offsetX and Y are relatives to the <svg> element
            x: Math.round(event.clientX - rect.left),
            y: Math.round(event.clientY - rect.top),
        },
        name,
        nameSource,
    };
}
function getEventTarget(event) {
    if (event.composed && (0, htmlDomUtils_1.isNodeShadowHost)(event.target) && typeof event.composedPath === 'function') {
        const composedPath = event.composedPath();
        if (composedPath.length > 0 && composedPath[0] instanceof Element) {
            return composedPath[0];
        }
    }
    return event.target;
}
function newClick(lifeCycle, actionTracker, getUserActivity, clickActionBase, startEvent) {
    const clickKey = (0, browser_core_1.generateUUID)();
    const startClocks = (0, browser_core_1.relativeToClocks)(startEvent.timeStamp);
    actionTracker.start(clickKey, startClocks, clickActionBase, { isChildEvent: isActionChildEvent_1.isActionChildEvent });
    let status = 0 /* ClickStatus.ONGOING */;
    let actionTrackerFinishedEvent;
    const frustrationTypes = [];
    const stopObservable = new browser_core_1.Observable();
    function stop(activityEndTime) {
        if (status !== 0 /* ClickStatus.ONGOING */) {
            return;
        }
        status = 1 /* ClickStatus.STOPPED */;
        actionTrackerFinishedEvent = activityEndTime
            ? actionTracker.stop(clickKey, (0, browser_core_1.timeStampToClocks)(activityEndTime))
            : actionTracker.discard(clickKey);
        stopObservable.notify();
    }
    return {
        event: startEvent,
        stop,
        stopObservable,
        get hasError() {
            var _a;
            const currentCounts = (_a = actionTrackerFinishedEvent === null || actionTrackerFinishedEvent === void 0 ? void 0 : actionTrackerFinishedEvent.counts) !== null && _a !== void 0 ? _a : actionTracker.getCounts(clickKey);
            return currentCounts ? currentCounts.errorCount > 0 : false;
        },
        get hasPageActivity() {
            return actionTrackerFinishedEvent && 'duration' in actionTrackerFinishedEvent;
        },
        getUserActivity,
        addFrustration: (frustrationType) => {
            frustrationTypes.push(frustrationType);
        },
        get startClocks() {
            return startClocks;
        },
        isStopped: () => status === 1 /* ClickStatus.STOPPED */ || status === 2 /* ClickStatus.FINALIZED */,
        clone: () => newClick(lifeCycle, actionTracker, getUserActivity, clickActionBase, startEvent),
        validate: (domEvents) => {
            stop();
            if (status !== 1 /* ClickStatus.STOPPED */) {
                return;
            }
            if (!actionTrackerFinishedEvent) {
                return;
            }
            const clickAction = {
                frustrationTypes,
                events: domEvents !== null && domEvents !== void 0 ? domEvents : [startEvent],
                event: startEvent,
                ...actionTrackerFinishedEvent,
                counts: actionTrackerFinishedEvent.counts, // This is needed to satisfy the type checker
            };
            lifeCycle.notify(0 /* LifeCycleEventType.AUTO_ACTION_COMPLETED */, clickAction);
            status = 2 /* ClickStatus.FINALIZED */;
        },
        discard: () => {
            stop();
            status = 2 /* ClickStatus.FINALIZED */;
        },
    };
}
function finalizeClicks(clicks, rageClick) {
    const { isRage } = (0, computeFrustration_1.computeFrustration)(clicks, rageClick);
    if (isRage) {
        clicks.forEach((click) => click.discard());
        rageClick.stop((0, browser_core_1.timeStampNow)());
        rageClick.validate(clicks.map((click) => click.event));
    }
    else {
        rageClick.discard();
        clicks.forEach((click) => click.validate());
    }
}
