// @datadog/browser-rum/esm/domain/record/serialization/insertionCursor.js
/** Returns an InsertionCursor which starts positioned at the root of the document. */
export function createRootInsertionCursor(nodeIds) {
    return createInsertionCursor(undefined, undefined, nodeIds);
}
/**
 * Returns an InsertionCursor which starts positioned in the child list of the given
 * parent node. If a next sibling is provided, the cursor points to the position
 * immediately before the next sibling; otherwise, the cursor points to the end of the
 * child list.
 */
export function createChildInsertionCursor(parentId, nextSiblingId, nodeIds) {
    return createInsertionCursor(parentId, nextSiblingId, nodeIds);
}
function createInsertionCursor(parentId, nextSiblingId, nodeIds) {
    let cursor = {
        container: undefined,
        parentId,
        previousSiblingId: undefined,
        nextSiblingId,
    };
    const computeInsertionPoint = (nodeId) => {
        if (cursor.previousSiblingId === nodeId - 1) {
            // Use an AppendAfterPreviousInsertionPoint. (i.e., 0)
            return 0;
        }
        if (cursor.nextSiblingId !== undefined) {
            // Use an InsertBeforeInsertionPoint. We identify the next sibling using a
            // negative integer indicating the difference between the new node's id and its next
            // sibling's id.
            return cursor.nextSiblingId - nodeId;
        }
        if (cursor.parentId !== undefined) {
            // Use an AppendChildInsertionPoint. We identify the parent node using a positive
            // integer indicating the difference between the new node's id and its parent's id.
            return nodeId - cursor.parentId;
        }
        // There's no parent. Use a RootInsertionPoint. (i.e., null)
        return null;
    };
    return {
        advance(node) {
            const nodeId = nodeIds.getOrInsert(node);
            const insertionPoint = computeInsertionPoint(nodeId);
            cursor.previousSiblingId = nodeId;
            return { nodeId, insertionPoint };
        },
        ascend() {
            if (cursor.container) {
                cursor = cursor.container;
            }
        },
        descend() {
            if (cursor.previousSiblingId !== undefined) {
                cursor = {
                    container: cursor,
                    parentId: cursor.previousSiblingId,
                    previousSiblingId: undefined,
                    nextSiblingId: undefined,
                };
            }
        },
    };
}
