// @datadog/browser-rum/esm/domain/record/serialization/serializationTransaction.js
import { elapsed, timeStampNow } from '@datadog/browser-core';
import { ChangeType, RecordType } from '../../../types';
import { createSerializationStats, updateSerializationStats } from './serializationStats';
import { createChangeEncoder } from './changeEncoder';
/**
 * Perform serialization within a transaction. At the end of the transaction, the
 * generated records and statistics will be emitted.
 */
export function serializeInTransaction(kind, emitRecord, emitStats, scope, serialize) {
    const records = [];
    const stats = createSerializationStats();
    const transaction = {
        add(record) {
            records.push(record);
        },
        addMetric(metric, value) {
            updateSerializationStats(stats, metric, value);
        },
        assignId(node) {
            const id = scope.nodeIds.getOrInsert(node);
            if (transaction.serializedNodeIds) {
                transaction.serializedNodeIds.add(id);
            }
            return id;
        },
        kind,
        scope,
    };
    const start = timeStampNow();
    serialize(transaction);
    updateSerializationStats(stats, 'serializationDuration', elapsed(start, timeStampNow()));
    for (const record of records) {
        emitRecord(record);
    }
    emitStats(stats);
}
export function serializeChangesInTransaction(kind, emitRecord, emitStats, scope, timestamp, serialize) {
    const encoder = createChangeEncoder(scope.stringIds);
    const stats = createSerializationStats();
    const transaction = {
        addMetric(metric, value) {
            updateSerializationStats(stats, metric, value);
        },
        addNode(...change) {
            encoder.add(ChangeType.AddNode, change);
        },
        addStyleSheet(rules, mediaList, disabled) {
            if (disabled) {
                encoder.add(ChangeType.AddStyleSheet, [rules, mediaList || [], disabled]);
            }
            else if (mediaList) {
                encoder.add(ChangeType.AddStyleSheet, [rules, mediaList]);
            }
            else {
                encoder.add(ChangeType.AddStyleSheet, [rules]);
            }
        },
        attachStyleSheets(nodeId, sheetIds) {
            const change = [nodeId];
            for (const sheetId of sheetIds) {
                change.push(sheetId);
            }
            encoder.add(ChangeType.AttachedStyleSheets, change);
        },
        setMediaPlaybackState(nodeId, state) {
            encoder.add(ChangeType.MediaPlaybackState, [nodeId, state]);
        },
        setScrollPosition(nodeId, x, y) {
            encoder.add(ChangeType.ScrollPosition, [nodeId, x, y]);
        },
        setSize(nodeId, width, height) {
            encoder.add(ChangeType.Size, [nodeId, width, height]);
        },
        kind,
        scope,
    };
    const start = timeStampNow();
    serialize(transaction);
    updateSerializationStats(stats, 'serializationDuration', elapsed(start, timeStampNow()));
    const changes = encoder.flush();
    if (changes.length > 0) {
        emitRecord({
            data: changes,
            type: RecordType.Change,
            timestamp,
        });
    }
    emitStats(stats);
}
