// @datadog/browser-rum/esm/domain/record/serialization/serializeFullSnapshot.js
import { getScrollX, getScrollY } from '@datadog/browser-rum-core';
import { RecordType } from '../../../types';
import { serializeNode } from './serializeNode';
import { serializeInTransaction } from './serializationTransaction';
export function serializeFullSnapshot(timestamp, kind, document, emitRecord, emitStats, scope) {
    serializeInTransaction(kind, emitRecord, emitStats, scope, (transaction) => {
        const defaultPrivacyLevel = transaction.scope.configuration.defaultPrivacyLevel;
        // We are sure that Documents are never ignored, so this function never returns null.
        const node = serializeNode(document, defaultPrivacyLevel, transaction);
        const record = {
            data: {
                node,
                initialOffset: {
                    left: getScrollX(),
                    top: getScrollY(),
                },
            },
            type: RecordType.FullSnapshot,
            timestamp,
        };
        transaction.add(record);
    });
}
