// @datadog/browser-rum/esm/domain/record/serialization/serializeNodeAsChange.js
import { reducePrivacyLevel, getNodeSelfPrivacyLevel, getTextContent, isNodeShadowRoot, forEachChildNodes, NodePrivacyLevel, PRIVACY_ATTR_NAME, PRIVACY_ATTR_VALUE_HIDDEN, getScrollX, getScrollY, isElementNode, } from '@datadog/browser-rum-core';
import { MediaInteractionType } from '../../../types';
import { serializeDOMAttributes, serializeVirtualAttributes } from './serializeAttributes';
export function serializeNodeAsChange(cursor, node, parentPrivacyLevel, transaction) {
    let privacyLevel;
    const selfPrivacyLevel = getNodeSelfPrivacyLevel(node);
    if (selfPrivacyLevel) {
        privacyLevel = reducePrivacyLevel(selfPrivacyLevel, parentPrivacyLevel);
    }
    else {
        privacyLevel = parentPrivacyLevel;
    }
    if (privacyLevel === NodePrivacyLevel.HIDDEN) {
        serializeHiddenNodePlaceholder(cursor, node, transaction);
        return;
    }
    // Totally ignore risky or unwanted elements. (e.g. <script>, some <link> and <meta> elements)
    if (privacyLevel === NodePrivacyLevel.IGNORE) {
        return;
    }
    switch (node.nodeType) {
        case node.CDATA_SECTION_NODE:
            serializeCDataNodeAsChange(cursor, node, transaction);
            break;
        case node.DOCUMENT_NODE:
            serializeDocumentNodeAsChange(cursor, node, transaction);
            break;
        case node.DOCUMENT_FRAGMENT_NODE:
            serializeDocumentFragmentNodeAsChange(cursor, node, transaction);
            break;
        case node.DOCUMENT_TYPE_NODE:
            serializeDocumentTypeNodeAsChange(cursor, node, transaction);
            break;
        case node.ELEMENT_NODE:
            serializeElementNodeAsChange(cursor, node, privacyLevel, transaction);
            break;
        case node.TEXT_NODE:
            serializeTextNodeAsChange(cursor, node, privacyLevel, transaction);
            break;
        default:
            return;
    }
    // If this node can't have children, we're done.
    switch (node.nodeType) {
        case node.CDATA_SECTION_NODE:
        case node.DOCUMENT_TYPE_NODE:
        case node.TEXT_NODE:
            return;
    }
    // Ignore the children of <style> elements; the CSS rules they contain are already
    // serialized as StyleSheetSnapshots.
    if (node.nodeName === 'STYLE') {
        return;
    }
    cursor.descend();
    forEachChildNodes(node, (childNode) => {
        serializeNodeAsChange(cursor, childNode, privacyLevel, transaction);
    });
    cursor.ascend();
}
function serializeDocumentNodeAsChange(cursor, document, transaction) {
    const { nodeId, insertionPoint } = cursor.advance(document);
    transaction.addNode(insertionPoint, '#document');
    transaction.setScrollPosition(nodeId, getScrollX(), getScrollY());
    serializeStyleSheetsAsChange(document.adoptedStyleSheets, nodeId, transaction);
}
function serializeDocumentFragmentNodeAsChange(cursor, documentFragment, transaction) {
    const { nodeId, insertionPoint } = cursor.advance(documentFragment);
    const isShadowRoot = isNodeShadowRoot(documentFragment);
    if (!isShadowRoot) {
        transaction.addNode(insertionPoint, '#document-fragment');
        return;
    }
    transaction.addNode(insertionPoint, '#shadow-root');
    transaction.scope.shadowRootsController.addShadowRoot(documentFragment, transaction.scope);
    serializeStyleSheetsAsChange(documentFragment.adoptedStyleSheets, nodeId, transaction);
}
function serializeDocumentTypeNodeAsChange(cursor, documentType, transaction) {
    const { insertionPoint } = cursor.advance(documentType);
    transaction.addNode(insertionPoint, '#doctype', documentType.name, documentType.publicId, documentType.systemId);
}
function serializeElementNodeAsChange(cursor, element, privacyLevel, transaction) {
    const { nodeId, insertionPoint } = cursor.advance(element);
    const domAttributes = Object.entries(serializeDOMAttributes(element, privacyLevel, transaction));
    transaction.addNode(insertionPoint, encodedElementName(element), ...domAttributes);
    const { _cssText: cssText, rr_mediaState: mediaState, rr_scrollLeft: scrollLeft, rr_scrollTop: scrollTop, } = serializeVirtualAttributes(element, privacyLevel, transaction);
    const linkOrStyle = element;
    if (cssText !== undefined && linkOrStyle.sheet) {
        const sheetId = transaction.scope.styleSheetIds.getOrInsert(linkOrStyle.sheet);
        transaction.addStyleSheet(cssText);
        transaction.attachStyleSheets(nodeId, [sheetId]);
    }
    if (mediaState === 'played') {
        transaction.setMediaPlaybackState(nodeId, MediaInteractionType.Play);
    }
    else if (mediaState === 'paused') {
        transaction.setMediaPlaybackState(nodeId, MediaInteractionType.Pause);
    }
    if (scrollLeft !== undefined || scrollTop !== undefined) {
        transaction.setScrollPosition(nodeId, scrollLeft || 0, scrollTop || 0);
    }
}
function serializeTextNodeAsChange(cursor, textNode, privacyLevel, transaction) {
    const textContent = getTextContent(textNode, privacyLevel);
    if (textContent === undefined) {
        return;
    }
    const { insertionPoint } = cursor.advance(textNode);
    transaction.addNode(insertionPoint, '#text', textContent);
}
function serializeCDataNodeAsChange(cursor, cdataNode, transaction) {
    const { insertionPoint } = cursor.advance(cdataNode);
    transaction.addNode(insertionPoint, '#cdata-section');
}
function serializeHiddenNodePlaceholder(cursor, node, transaction) {
    // We only generate placeholders for element nodes; other hidden nodes are simply not
    // serialized.
    if (!isElementNode(node)) {
        return;
    }
    const { nodeId, insertionPoint } = cursor.advance(node);
    transaction.addNode(insertionPoint, encodedElementName(node), [PRIVACY_ATTR_NAME, PRIVACY_ATTR_VALUE_HIDDEN]);
    const { width, height } = node.getBoundingClientRect();
    transaction.setSize(nodeId, width, height);
}
function serializeStyleSheetsAsChange(sheets, nodeId, transaction) {
    if (!sheets || sheets.length === 0) {
        return undefined;
    }
    transaction.attachStyleSheets(nodeId, sheets.map((sheet) => serializeStyleSheetAsChange(sheet, transaction)));
}
function serializeStyleSheetAsChange(sheet, transaction) {
    const rules = Array.from(sheet.cssRules || sheet.rules, (rule) => rule.cssText);
    const mediaList = sheet.media.length > 0 ? Array.from(sheet.media) : undefined;
    transaction.addMetric('cssText', rules.reduce((totalLength, rule) => totalLength + rule.length, 0));
    transaction.addStyleSheet(rules, mediaList, sheet.disabled);
    return transaction.scope.styleSheetIds.getOrInsert(sheet);
}
function encodedElementName(element) {
    const nodeName = element.nodeName;
    if (isSVGElement(element)) {
        return `svg>${nodeName}`;
    }
    return nodeName;
}
function isSVGElement(element) {
    return element.namespaceURI === 'http://www.w3.org/2000/svg';
}
