// @datadog/browser-rum/cjs/domain/profiling/longTaskHistory.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LONG_TASK_ID_HISTORY_EXPIRE_DELAY = void 0;
exports.createLongTaskHistory = createLongTaskHistory;
const browser_core_1 = require("@datadog/browser-core");
const browser_rum_core_1 = require("@datadog/browser-rum-core");
exports.LONG_TASK_ID_HISTORY_EXPIRE_DELAY = browser_core_1.SESSION_TIME_OUT_DELAY;
function createLongTaskHistory(lifeCycle) {
    const history = (0, browser_core_1.createValueHistory)({
        expireDelay: exports.LONG_TASK_ID_HISTORY_EXPIRE_DELAY,
    });
    lifeCycle.subscribe(12 /* LifeCycleEventType.RAW_RUM_EVENT_COLLECTED */, ({ rawRumEvent, startClocks, duration }) => {
        if (rawRumEvent.type === 'long_task') {
            history
                .add({
                id: rawRumEvent.long_task.id,
                startClocks,
                duration: duration,
                entryType: rawRumEvent.long_task.entry_type === browser_rum_core_1.RumLongTaskEntryType.LONG_TASK
                    ? browser_rum_core_1.RumPerformanceEntryType.LONG_TASK
                    : browser_rum_core_1.RumPerformanceEntryType.LONG_ANIMATION_FRAME,
            }, startClocks.relative)
                .close((0, browser_core_1.addDuration)(startClocks.relative, duration));
        }
    });
    return history;
}
