// @datadog/browser-rum-core/esm/domain/resource/trackManualResources.js
import { clocksNow, elapsed, ResourceType as ResourceTypeEnum, toServerDuration } from '@datadog/browser-core';
import { RumEventType } from '../../rawRumEvent.types';
import { sanitizeIfLongDataUrl } from './resourceUtils';
export function trackManualResources(lifeCycle, resourceTracker) {
    function emitResource(id, startClocks, data, statusCode, endClocks) {
        const duration = endClocks ? elapsed(startClocks.relative, endClocks.relative) : undefined;
        const rawRumEvent = {
            date: startClocks.timeStamp,
            type: RumEventType.RESOURCE,
            resource: {
                id,
                type: data.type || ResourceTypeEnum.OTHER,
                url: sanitizeIfLongDataUrl(data.url),
                duration: duration !== undefined ? toServerDuration(duration) : undefined,
                method: data.method,
                status_code: statusCode,
            },
            _dd: {},
            context: data.context,
        };
        lifeCycle.notify(12 /* LifeCycleEventType.RAW_RUM_EVENT_COLLECTED */, {
            rawRumEvent,
            startClocks,
            duration,
            domainContext: { isManual: true },
        });
    }
    function startManualResource(url, options = {}, startClocks = clocksNow()) {
        var _a;
        const lookupKey = (_a = options.resourceKey) !== null && _a !== void 0 ? _a : url;
        resourceTracker.start(lookupKey, startClocks, {
            url,
            ...options,
        });
    }
    function stopManualResource(url, options = {}, stopClocks = clocksNow()) {
        var _a;
        const lookupKey = (_a = options.resourceKey) !== null && _a !== void 0 ? _a : url;
        const stopped = resourceTracker.stop(lookupKey, stopClocks, {
            context: options.context,
            type: options.type,
        });
        if (!stopped) {
            return;
        }
        emitResource(stopped.id, stopped.startClocks, stopped, options.statusCode, stopClocks);
    }
    return {
        startResource: startManualResource,
        stopResource: stopManualResource,
    };
}
