// @datadog/browser-rum/esm/domain/record/trackers/trackViewportResize.js
import { throttle, addEventListeners, timeStampNow, noop } from '@datadog/browser-core';
import { initViewportObservable } from '@datadog/browser-rum-core';
import { IncrementalSource, RecordType } from '../../../types';
import { getVisualViewport } from '../viewports';
import { assembleIncrementalSnapshot } from '../assembly';
const VISUAL_VIEWPORT_OBSERVER_THRESHOLD = 200;
export function trackViewportResize(emitRecord, scope) {
    const viewportResizeSubscription = initViewportObservable(scope.configuration).subscribe((data) => {
        emitRecord(assembleIncrementalSnapshot(IncrementalSource.ViewportResize, data));
    });
    return {
        stop: () => {
            viewportResizeSubscription.unsubscribe();
        },
    };
}
export function trackVisualViewportResize(emitRecord, scope) {
    const visualViewport = window.visualViewport;
    if (!visualViewport) {
        return { stop: noop };
    }
    const { throttled: updateDimension, cancel: cancelThrottle } = throttle(() => {
        emitRecord({
            data: getVisualViewport(visualViewport),
            type: RecordType.VisualViewport,
            timestamp: timeStampNow(),
        });
    }, VISUAL_VIEWPORT_OBSERVER_THRESHOLD, {
        trailing: false,
    });
    const { stop: removeListener } = addEventListeners(scope.configuration, visualViewport, ["resize" /* DOM_EVENT.RESIZE */, "scroll" /* DOM_EVENT.SCROLL */], updateDimension, {
        capture: true,
        passive: true,
    });
    return {
        stop: () => {
            removeListener();
            cancelThrottle();
        },
    };
}
