// @datadog/browser-rum/esm/domain/record/trackers/trackMutation.js
import { monitor, noop, timeStampNow } from '@datadog/browser-core';
import { getMutationObserverConstructor } from '@datadog/browser-rum-core';
import { createMutationBatch } from '../mutationBatch';
import { serializeMutations } from '../serialization';
/**
 * Buffers and aggregate mutations generated by a MutationObserver into MutationPayload
 */
export function trackMutation(target, emitRecord, emitStats, scope, serialize = defaultSerializeMutationsCallback()) {
    const MutationObserver = getMutationObserverConstructor();
    if (!MutationObserver) {
        return { stop: noop, flush: noop };
    }
    const mutationBatch = createMutationBatch((mutations) => {
        serialize(timeStampNow(), mutations.concat(observer.takeRecords()), emitRecord, emitStats, scope);
    });
    const observer = new MutationObserver(monitor(mutationBatch.addMutations));
    observer.observe(target, {
        attributeOldValue: true,
        attributes: true,
        characterData: true,
        characterDataOldValue: true,
        childList: true,
        subtree: true,
    });
    return {
        stop: () => {
            observer.disconnect();
            mutationBatch.stop();
        },
        flush: () => {
            mutationBatch.flush();
        },
    };
}
function defaultSerializeMutationsCallback() {
    return serializeMutations;
}
