// @datadog/browser-rum/cjs/domain/record/serialization/serializationTransaction.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.serializeInTransaction = serializeInTransaction;
exports.serializeChangesInTransaction = serializeChangesInTransaction;
const browser_core_1 = require("@datadog/browser-core");
const types_1 = require("../../../types");
const serializationStats_1 = require("./serializationStats");
const changeEncoder_1 = require("./changeEncoder");
/**
 * Perform serialization within a transaction. At the end of the transaction, the
 * generated records and statistics will be emitted.
 */
function serializeInTransaction(kind, emitRecord, emitStats, scope, serialize) {
    const records = [];
    const stats = (0, serializationStats_1.createSerializationStats)();
    const transaction = {
        add(record) {
            records.push(record);
        },
        addMetric(metric, value) {
            (0, serializationStats_1.updateSerializationStats)(stats, metric, value);
        },
        assignId(node) {
            const id = scope.nodeIds.getOrInsert(node);
            if (transaction.serializedNodeIds) {
                transaction.serializedNodeIds.add(id);
            }
            return id;
        },
        kind,
        scope,
    };
    const start = (0, browser_core_1.timeStampNow)();
    serialize(transaction);
    (0, serializationStats_1.updateSerializationStats)(stats, 'serializationDuration', (0, browser_core_1.elapsed)(start, (0, browser_core_1.timeStampNow)()));
    for (const record of records) {
        emitRecord(record);
    }
    emitStats(stats);
}
function serializeChangesInTransaction(kind, emitRecord, emitStats, scope, timestamp, serialize) {
    const encoder = (0, changeEncoder_1.createChangeEncoder)(scope.stringIds);
    const stats = (0, serializationStats_1.createSerializationStats)();
    const transaction = {
        addMetric(metric, value) {
            (0, serializationStats_1.updateSerializationStats)(stats, metric, value);
        },
        addNode(...change) {
            encoder.add(types_1.ChangeType.AddNode, change);
        },
        addStyleSheet(rules, mediaList, disabled) {
            if (disabled) {
                encoder.add(types_1.ChangeType.AddStyleSheet, [rules, mediaList || [], disabled]);
            }
            else if (mediaList) {
                encoder.add(types_1.ChangeType.AddStyleSheet, [rules, mediaList]);
            }
            else {
                encoder.add(types_1.ChangeType.AddStyleSheet, [rules]);
            }
        },
        attachStyleSheets(nodeId, sheetIds) {
            const change = [nodeId];
            for (const sheetId of sheetIds) {
                change.push(sheetId);
            }
            encoder.add(types_1.ChangeType.AttachedStyleSheets, change);
        },
        setMediaPlaybackState(nodeId, state) {
            encoder.add(types_1.ChangeType.MediaPlaybackState, [nodeId, state]);
        },
        setScrollPosition(nodeId, x, y) {
            encoder.add(types_1.ChangeType.ScrollPosition, [nodeId, x, y]);
        },
        setSize(nodeId, width, height) {
            encoder.add(types_1.ChangeType.Size, [nodeId, width, height]);
        },
        kind,
        scope,
    };
    const start = (0, browser_core_1.timeStampNow)();
    serialize(transaction);
    (0, serializationStats_1.updateSerializationStats)(stats, 'serializationDuration', (0, browser_core_1.elapsed)(start, (0, browser_core_1.timeStampNow)()));
    const changes = encoder.flush();
    if (changes.length > 0) {
        emitRecord({
            data: changes,
            type: types_1.RecordType.Change,
            timestamp,
        });
    }
    emitStats(stats);
}
