// @datadog/browser-rum/cjs/domain/record/serialization/serializeFullSnapshot.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.serializeFullSnapshot = serializeFullSnapshot;
const browser_rum_core_1 = require("@datadog/browser-rum-core");
const types_1 = require("../../../types");
const serializeNode_1 = require("./serializeNode");
const serializationTransaction_1 = require("./serializationTransaction");
function serializeFullSnapshot(timestamp, kind, document, emitRecord, emitStats, scope) {
    (0, serializationTransaction_1.serializeInTransaction)(kind, emitRecord, emitStats, scope, (transaction) => {
        const defaultPrivacyLevel = transaction.scope.configuration.defaultPrivacyLevel;
        // We are sure that Documents are never ignored, so this function never returns null.
        const node = (0, serializeNode_1.serializeNode)(document, defaultPrivacyLevel, transaction);
        const record = {
            data: {
                node,
                initialOffset: {
                    left: (0, browser_rum_core_1.getScrollX)(),
                    top: (0, browser_rum_core_1.getScrollY)(),
                },
            },
            type: types_1.RecordType.FullSnapshot,
            timestamp,
        };
        transaction.add(record);
    });
}
