// @datadog/browser-rum-react/cjs/domain/reactPlugin.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.reactPlugin = reactPlugin;
exports.onRumInit = onRumInit;
exports.onRumStart = onRumStart;
exports.resetReactPlugin = resetReactPlugin;
let globalPublicApi;
let globalConfiguration;
let globalAddEvent;
const onRumInitSubscribers = [];
const onRumStartSubscribers = [];
/**
 * React plugin constructor.
 *
 * @category Main
 * @example
 * ```ts
 * import { datadogRum } from '@datadog/browser-rum'
 * import { reactPlugin } from '@datadog/browser-rum-react'
 *
 * datadogRum.init({
 *   applicationId: '<DATADOG_APPLICATION_ID>',
 *   clientToken: '<DATADOG_CLIENT_TOKEN>',
 *   site: '<DATADOG_SITE>',
 *   plugins: [reactPlugin()],
 *   // ...
 * })
 * ```
 */
function reactPlugin(configuration = {}) {
    return {
        name: 'react',
        onInit({ publicApi, initConfiguration }) {
            globalPublicApi = publicApi;
            globalConfiguration = configuration;
            for (const subscriber of onRumInitSubscribers) {
                subscriber(globalConfiguration, globalPublicApi);
            }
            if (configuration.router) {
                initConfiguration.trackViewsManually = true;
            }
        },
        onRumStart({ addEvent }) {
            globalAddEvent = addEvent;
            for (const subscriber of onRumStartSubscribers) {
                if (addEvent) {
                    subscriber(addEvent);
                }
            }
        },
        getConfigurationTelemetry() {
            return { router: !!configuration.router };
        },
    };
}
function onRumInit(callback) {
    if (globalConfiguration && globalPublicApi) {
        callback(globalConfiguration, globalPublicApi);
    }
    else {
        onRumInitSubscribers.push(callback);
    }
}
function onRumStart(callback) {
    if (globalAddEvent) {
        callback(globalAddEvent);
    }
    else {
        onRumStartSubscribers.push(callback);
    }
}
function resetReactPlugin() {
    globalPublicApi = undefined;
    globalConfiguration = undefined;
    globalAddEvent = undefined;
    onRumInitSubscribers.length = 0;
    onRumStartSubscribers.length = 0;
}
