// @datadog/browser-core/cjs/domain/synthetics/syntheticsWorkerValues.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SYNTHETICS_CONTEXT_COOKIE_NAME = exports.SYNTHETICS_INJECTS_RUM_COOKIE_NAME = exports.SYNTHETICS_RESULT_ID_COOKIE_NAME = exports.SYNTHETICS_TEST_ID_COOKIE_NAME = void 0;
exports.willSyntheticsInjectRum = willSyntheticsInjectRum;
exports.getSyntheticsContext = getSyntheticsContext;
exports.isSyntheticsTest = isSyntheticsTest;
const cookie_1 = require("../../browser/cookie");
const globalObject_1 = require("../../tools/globalObject");
const objectUtils_1 = require("../../tools/utils/objectUtils");
const cookieNamePrefix = 'datadog-synthetics-';
exports.SYNTHETICS_TEST_ID_COOKIE_NAME = `${cookieNamePrefix}public-id`;
exports.SYNTHETICS_RESULT_ID_COOKIE_NAME = `${cookieNamePrefix}result-id`;
exports.SYNTHETICS_INJECTS_RUM_COOKIE_NAME = `${cookieNamePrefix}injects-rum`;
exports.SYNTHETICS_CONTEXT_COOKIE_NAME = `${cookieNamePrefix}rum-context`;
function willSyntheticsInjectRum() {
    if (globalObject_1.isWorkerEnvironment) {
        // We don't expect to run synthetics tests in a worker environment
        return false;
    }
    return Boolean(globalObject_1.globalObject._DATADOG_SYNTHETICS_INJECTS_RUM || (0, cookie_1.getInitCookie)(exports.SYNTHETICS_INJECTS_RUM_COOKIE_NAME));
}
function getSyntheticsContext() {
    const raw = getRawSyntheticsContext();
    return isValidSyntheticsContext(raw) ? raw : undefined;
}
function isSyntheticsTest() {
    return Boolean(getSyntheticsContext());
}
function getRawSyntheticsContext() {
    const rawGlobal = globalObject_1.globalObject._DATADOG_SYNTHETICS_RUM_CONTEXT;
    if (rawGlobal) {
        return rawGlobal;
    }
    const rawCookie = (0, cookie_1.getInitCookie)(exports.SYNTHETICS_CONTEXT_COOKIE_NAME);
    if (rawCookie) {
        return (0, objectUtils_1.tryJsonParse)(decodeURIComponent(rawCookie));
    }
    return {
        test_id: window._DATADOG_SYNTHETICS_PUBLIC_ID || (0, cookie_1.getInitCookie)(exports.SYNTHETICS_TEST_ID_COOKIE_NAME),
        result_id: window._DATADOG_SYNTHETICS_RESULT_ID || (0, cookie_1.getInitCookie)(exports.SYNTHETICS_RESULT_ID_COOKIE_NAME),
    };
}
function isValidSyntheticsContext(value) {
    return (typeof value === 'object' &&
        value !== null &&
        typeof value.test_id === 'string' &&
        typeof value.result_id === 'string');
}
