// @datadog/browser-rum-react/cjs/entries/reactRouterV6.js
"use strict";
/**
 * React Router v6 integration.
 *
 * @packageDocumentation
 * @example
 * ```ts
 * import { RouterProvider } from 'react-router-dom'
 * import { datadogRum } from '@datadog/browser-rum'
 * import { reactPlugin } from '@datadog/browser-rum-react'
 *
 * // ⚠️ Use "createBrowserRouter" from `@datadog/browser-rum-react/react-router-v6` instead of `react-router-dom`
 * import { createBrowserRouter } from '@datadog/browser-rum-react/react-router-v6'
 *
 * datadogRum.init({
 *   applicationId: '<DATADOG_APPLICATION_ID>',
 *   clientToken: '<DATADOG_CLIENT_TOKEN>',
 *   plugins: [reactPlugin({ router: true })],
 *   // ...
 * })
 *
 * const router = createBrowserRouter([
 *   {
 *     path: '/',
 *     element: <Root />,
 *     // ...
 *   },
 * ])
 *
 * ReactDOM.createRoot(document.getElementById('root')).render(<RouterProvider router={router} />)
 * ```
 */
/* eslint-disable local-rules/disallow-side-effects */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Routes = exports.useRoutes = exports.createMemoryRouter = exports.createHashRouter = exports.createBrowserRouter = void 0;
const react_router_dom_1 = require("react-router-dom");
const reactRouter_1 = require("../domain/reactRouter");
/**
 * Use this function in place of `react-router-dom` `createBrowserRouter`. Every time a route is
 * rendered, a new RUM view is created.
 *
 * @see https://reactrouter.com/v6/routers/create-browser-router
 */
exports.createBrowserRouter = (0, reactRouter_1.wrapCreateRouter)(react_router_dom_1.createBrowserRouter);
/**
 * Use this function in place of `react-router-dom` `createHashRouter`. Every time a route is
 * rendered, a new RUM view is created.
 *
 * @see https://reactrouter.com/v6/routers/create-hash-router
 */
exports.createHashRouter = (0, reactRouter_1.wrapCreateRouter)(react_router_dom_1.createHashRouter);
/**
 * Use this function in place of `react-router-dom` `createMemoryRouter`. Every time a route is
 * rendered, a new RUM view is created.
 *
 * @see https://reactrouter.com/v6/routers/create-memory-router
 */
exports.createMemoryRouter = (0, reactRouter_1.wrapCreateRouter)(react_router_dom_1.createMemoryRouter);
/**
 * Use this hook in place of `react-router-dom` `useRoutes`. Every time a route is rendered, a new
 * RUM view is created.
 *
 * @see https://reactrouter.com/docs/en/v6/hooks/use-routes
 */
exports.useRoutes = (0, reactRouter_1.wrapUseRoutes)({
    useRoutes: react_router_dom_1.useRoutes,
    useLocation: react_router_dom_1.useLocation,
    matchRoutes: react_router_dom_1.matchRoutes,
});
/**
 * Use this component in place of `react-router-dom` `Routes`. Every time a route is rendered, a new
 * RUM view is created.
 *
 * @see https://reactrouter.com/docs/en/v6/components/routes
 */
exports.Routes = (0, reactRouter_1.createRoutesComponent)(exports.useRoutes, react_router_dom_1.createRoutesFromChildren);
