// @datadog/browser-rum-core/cjs/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createFormDataTransport = exports.isSampled = exports.createHooks = exports.RumPerformanceEntryType = exports.supportPerformanceTimingEvent = exports.sanitizeIfLongDataUrl = exports.getSessionReplayUrl = exports.STABLE_ATTRIBUTES = exports.DEFAULT_PROGRAMMATIC_ACTION_NAME_ATTRIBUTE = exports.getScrollY = exports.getScrollX = exports.getViewportDimension = exports.initViewportObservable = exports.getMutationObserverConstructor = exports.startViewHistory = exports.LifeCycle = exports.startRum = exports.RumLongTaskEntryType = exports.FrustrationType = exports.RumEventType = exports.ActionType = exports.makeRumPublicApi = void 0;
var rumPublicApi_1 = require("./boot/rumPublicApi");
Object.defineProperty(exports, "makeRumPublicApi", { enumerable: true, get: function () { return rumPublicApi_1.makeRumPublicApi; } });
var rawRumEvent_types_1 = require("./rawRumEvent.types");
Object.defineProperty(exports, "ActionType", { enumerable: true, get: function () { return rawRumEvent_types_1.ActionType; } });
Object.defineProperty(exports, "RumEventType", { enumerable: true, get: function () { return rawRumEvent_types_1.RumEventType; } });
Object.defineProperty(exports, "FrustrationType", { enumerable: true, get: function () { return rawRumEvent_types_1.FrustrationType; } });
Object.defineProperty(exports, "RumLongTaskEntryType", { enumerable: true, get: function () { return rawRumEvent_types_1.RumLongTaskEntryType; } });
var startRum_1 = require("./boot/startRum");
Object.defineProperty(exports, "startRum", { enumerable: true, get: function () { return startRum_1.startRum; } });
var lifeCycle_1 = require("./domain/lifeCycle");
Object.defineProperty(exports, "LifeCycle", { enumerable: true, get: function () { return lifeCycle_1.LifeCycle; } });
var viewHistory_1 = require("./domain/contexts/viewHistory");
Object.defineProperty(exports, "startViewHistory", { enumerable: true, get: function () { return viewHistory_1.startViewHistory; } });
var domMutationObservable_1 = require("./browser/domMutationObservable");
Object.defineProperty(exports, "getMutationObserverConstructor", { enumerable: true, get: function () { return domMutationObservable_1.getMutationObserverConstructor; } });
var viewportObservable_1 = require("./browser/viewportObservable");
Object.defineProperty(exports, "initViewportObservable", { enumerable: true, get: function () { return viewportObservable_1.initViewportObservable; } });
Object.defineProperty(exports, "getViewportDimension", { enumerable: true, get: function () { return viewportObservable_1.getViewportDimension; } });
var scroll_1 = require("./browser/scroll");
Object.defineProperty(exports, "getScrollX", { enumerable: true, get: function () { return scroll_1.getScrollX; } });
Object.defineProperty(exports, "getScrollY", { enumerable: true, get: function () { return scroll_1.getScrollY; } });
var actionNameConstants_1 = require("./domain/action/actionNameConstants");
Object.defineProperty(exports, "DEFAULT_PROGRAMMATIC_ACTION_NAME_ATTRIBUTE", { enumerable: true, get: function () { return actionNameConstants_1.DEFAULT_PROGRAMMATIC_ACTION_NAME_ATTRIBUTE; } });
var getSelectorFromElement_1 = require("./domain/getSelectorFromElement");
Object.defineProperty(exports, "STABLE_ATTRIBUTES", { enumerable: true, get: function () { return getSelectorFromElement_1.STABLE_ATTRIBUTES; } });
__exportStar(require("./browser/htmlDomUtils"), exports);
var getSessionReplayUrl_1 = require("./domain/getSessionReplayUrl");
Object.defineProperty(exports, "getSessionReplayUrl", { enumerable: true, get: function () { return getSessionReplayUrl_1.getSessionReplayUrl; } });
var resourceUtils_1 = require("./domain/resource/resourceUtils");
Object.defineProperty(exports, "sanitizeIfLongDataUrl", { enumerable: true, get: function () { return resourceUtils_1.sanitizeIfLongDataUrl; } });
__exportStar(require("./domain/privacy"), exports);
__exportStar(require("./domain/privacyConstants"), exports);
var performanceObservable_1 = require("./browser/performanceObservable");
Object.defineProperty(exports, "supportPerformanceTimingEvent", { enumerable: true, get: function () { return performanceObservable_1.supportPerformanceTimingEvent; } });
var performanceObservable_2 = require("./browser/performanceObservable");
Object.defineProperty(exports, "RumPerformanceEntryType", { enumerable: true, get: function () { return performanceObservable_2.RumPerformanceEntryType; } });
var hooks_1 = require("./domain/hooks");
Object.defineProperty(exports, "createHooks", { enumerable: true, get: function () { return hooks_1.createHooks; } });
var sampler_1 = require("./domain/sampler/sampler");
Object.defineProperty(exports, "isSampled", { enumerable: true, get: function () { return sampler_1.isSampled; } });
var formDataTransport_1 = require("./transport/formDataTransport");
Object.defineProperty(exports, "createFormDataTransport", { enumerable: true, get: function () { return formDataTransport_1.createFormDataTransport; } });
