// @datadog/browser-rum-core/cjs/domain/view/viewMetrics/trackNavigationTimings.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.trackNavigationTimings = trackNavigationTimings;
const browser_core_1 = require("@datadog/browser-core");
const performanceUtils_1 = require("../../../browser/performanceUtils");
function trackNavigationTimings(configuration, callback) {
    return waitAfterLoadEvent(configuration, () => {
        const entry = (0, browser_core_1.mockable)(performanceUtils_1.getNavigationEntry)();
        if (!isIncompleteNavigation(entry)) {
            callback(processNavigationEntry(entry));
        }
    });
}
function processNavigationEntry(entry) {
    return {
        domComplete: entry.domComplete,
        domContentLoaded: entry.domContentLoadedEventEnd,
        domInteractive: entry.domInteractive,
        loadEvent: entry.loadEventEnd,
        firstByte: (0, performanceUtils_1.sanitizeFirstByte)(entry),
    };
}
function isIncompleteNavigation(entry) {
    return entry.loadEventEnd <= 0;
}
function waitAfterLoadEvent(configuration, callback) {
    let timeoutId;
    const { stop: stopOnReadyState } = (0, browser_core_1.runOnReadyState)(configuration, 'complete', () => {
        // Invoke the callback a bit after the actual load event, so the "loadEventEnd" timing is accurate
        timeoutId = (0, browser_core_1.setTimeout)(() => callback());
    });
    return {
        stop: () => {
            stopOnReadyState();
            (0, browser_core_1.clearTimeout)(timeoutId);
        },
    };
}
