// @datadog/browser-rum/cjs/domain/record/trackers/trackMutation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.trackMutation = trackMutation;
const browser_core_1 = require("@datadog/browser-core");
const browser_rum_core_1 = require("@datadog/browser-rum-core");
const mutationBatch_1 = require("../mutationBatch");
const serialization_1 = require("../serialization");
/**
 * Buffers and aggregate mutations generated by a MutationObserver into MutationPayload
 */
function trackMutation(target, emitRecord, emitStats, scope, serialize = defaultSerializeMutationsCallback()) {
    const MutationObserver = (0, browser_rum_core_1.getMutationObserverConstructor)();
    if (!MutationObserver) {
        return { stop: browser_core_1.noop, flush: browser_core_1.noop };
    }
    const mutationBatch = (0, mutationBatch_1.createMutationBatch)((mutations) => {
        serialize((0, browser_core_1.timeStampNow)(), mutations.concat(observer.takeRecords()), emitRecord, emitStats, scope);
    });
    const observer = new MutationObserver((0, browser_core_1.monitor)(mutationBatch.addMutations));
    observer.observe(target, {
        attributeOldValue: true,
        attributes: true,
        characterData: true,
        characterDataOldValue: true,
        childList: true,
        subtree: true,
    });
    return {
        stop: () => {
            observer.disconnect();
            mutationBatch.stop();
        },
        flush: () => {
            mutationBatch.flush();
        },
    };
}
function defaultSerializeMutationsCallback() {
    return serialization_1.serializeMutations;
}
