// @datadog/browser-rum-react/cjs/domain/performance/reactComponentTracker.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.UNSTABLE_ReactComponentTracker = void 0;
const React = __importStar(require("react"));
const timer_1 = require("./timer");
const addDurationVital_1 = require("./addDurationVital");
/**
 * Track the performance of a React component.
 *
 * @experimental
 */
// eslint-disable-next-line
const UNSTABLE_ReactComponentTracker = ({ name: componentName, children, }) => {
    const isFirstRender = React.useRef(true);
    const renderTimer = (0, timer_1.createTimer)();
    const effectTimer = (0, timer_1.createTimer)();
    const layoutEffectTimer = (0, timer_1.createTimer)();
    const onEffectEnd = () => {
        var _a, _b, _c;
        const renderDuration = (_a = renderTimer.getDuration()) !== null && _a !== void 0 ? _a : 0;
        const effectDuration = (_b = effectTimer.getDuration()) !== null && _b !== void 0 ? _b : 0;
        const layoutEffectDuration = (_c = layoutEffectTimer.getDuration()) !== null && _c !== void 0 ? _c : 0;
        const totalRenderTime = renderDuration + effectDuration + layoutEffectDuration;
        (0, addDurationVital_1.addDurationVital)('reactComponentRender', {
            description: componentName,
            startTime: renderTimer.getStartTime(), // note: renderTimer should have been started at this point, so getStartTime should not return undefined
            duration: totalRenderTime,
            context: {
                is_first_render: isFirstRender.current,
                render_phase_duration: renderDuration,
                effect_phase_duration: effectDuration,
                layout_effect_phase_duration: layoutEffectDuration,
                framework: 'react',
            },
        });
        isFirstRender.current = false;
    };
    // In react, children are rendered sequentially in the order they are defined. that's why we can
    // measure perf timings of a component by starting recordings in the component above and stopping
    // them in the component below.
    return (React.createElement(React.Fragment, null,
        React.createElement(LifeCycle, { onRender: renderTimer.startTimer, onLayoutEffect: layoutEffectTimer.startTimer, onEffect: effectTimer.startTimer }),
        children,
        React.createElement(LifeCycle, { onRender: renderTimer.stopTimer, onLayoutEffect: layoutEffectTimer.stopTimer, onEffect: () => {
                effectTimer.stopTimer();
                onEffectEnd();
            } })));
};
exports.UNSTABLE_ReactComponentTracker = UNSTABLE_ReactComponentTracker;
function LifeCycle({ onRender, onLayoutEffect, onEffect, }) {
    onRender();
    React.useLayoutEffect(onLayoutEffect);
    React.useEffect(onEffect);
    return null;
}
