// @datadog/browser-rum-react/cjs/domain/error/addReactError.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.addReactError = addReactError;
const browser_core_1 = require("@datadog/browser-core");
const browser_rum_core_1 = require("@datadog/browser-rum-core");
const reactPlugin_1 = require("../reactPlugin");
/**
 * Add a React error to the RUM session.
 *
 * @category Error
 * @example
 * ```ts
 * import { createRoot } from 'react-dom/client'
 * import { datadogRum } from '@datadog/browser-rum'
 * import { addReactError } from '@datadog/browser-rum-react'
 *
 * const container = document.getElementById('root')
 * const root = createRoot(container, {
 *   onUncaughtError: (error, errorInfo) => {
 *     // Report uncaught errors to Datadog
 *     addReactError(error, errorInfo)
 *   }
 * })
 * // ...
 * ```
 */
function addReactError(error, info) {
    const handlingStack = (0, browser_core_1.createHandlingStack)('react error');
    const startClocks = (0, browser_core_1.clocksNow)();
    (0, reactPlugin_1.onRumStart)((addEvent) => {
        (0, browser_core_1.callMonitored)(() => {
            var _a;
            const rawError = (0, browser_core_1.computeRawError)({
                originalError: error,
                handlingStack,
                componentStack: (_a = info.componentStack) !== null && _a !== void 0 ? _a : undefined,
                startClocks,
                source: browser_core_1.ErrorSource.CUSTOM,
                handling: "handled" /* ErrorHandling.HANDLED */,
                nonErrorPrefix: "Provided" /* NonErrorPrefix.PROVIDED */,
            });
            addEvent(startClocks.relative, {
                type: browser_rum_core_1.RumEventType.ERROR,
                date: rawError.startClocks.timeStamp,
                error: {
                    id: (0, browser_core_1.generateUUID)(),
                    message: rawError.message,
                    source: rawError.source,
                    stack: rawError.stack,
                    handling_stack: rawError.handlingStack,
                    component_stack: rawError.componentStack,
                    type: rawError.type,
                    handling: rawError.handling,
                    causes: rawError.causes,
                    source_type: 'browser',
                    csp: rawError.csp,
                },
                context: { framework: 'react', ...rawError.context },
            }, {
                error: rawError.originalError,
                handlingStack: rawError.handlingStack,
            });
        });
    });
}
