// @datadog/browser-rum-react/cjs/domain/error/errorBoundary.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ErrorBoundary = void 0;
const react_1 = __importDefault(require("react"));
const addReactError_1 = require("./addReactError");
const INITIAL_STATE = { didCatch: false, error: null };
/**
 * ErrorBoundary component to report React errors to Datadog.
 *
 * For more advanced error handling, you can use the {@link addReactError} function.
 *
 * @category Error
 * @example
 * ```ts
 * import { ErrorBoundary } from '@datadog/browser-rum-react'
 *
 * <ErrorBoundary fallback={() => null}>
 *   <Component />
 * </ErrorBoundary>
 * ```
 */
// eslint-disable-next-line no-restricted-syntax
class ErrorBoundary extends react_1.default.Component {
    constructor(props) {
        super(props);
        this.resetError = () => {
            this.setState(INITIAL_STATE);
        };
        this.state = INITIAL_STATE;
    }
    static getDerivedStateFromError(error) {
        return { didCatch: true, error };
    }
    componentDidCatch(error, errorInfo) {
        (0, addReactError_1.addReactError)(error, errorInfo);
    }
    render() {
        if (this.state.didCatch) {
            return react_1.default.createElement(this.props.fallback, {
                error: this.state.error,
                resetError: this.resetError,
            });
        }
        return this.props.children;
    }
}
exports.ErrorBoundary = ErrorBoundary;
