// @datadog/browser-rum-core/cjs/domain/vital/vitalCollection.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createCustomVitalsState = createCustomVitalsState;
exports.startVitalCollection = startVitalCollection;
exports.startDurationVital = startDurationVital;
exports.stopDurationVital = stopDurationVital;
const browser_core_1 = require("@datadog/browser-core");
const rawRumEvent_types_1 = require("../../rawRumEvent.types");
function createCustomVitalsState() {
    const vitalsByName = new Map();
    const vitalsByReference = new WeakMap();
    return { vitalsByName, vitalsByReference };
}
function startVitalCollection(lifeCycle, pageStateHistory, customVitalsState) {
    function isValid(vital) {
        return !pageStateHistory.wasInPageStateDuringPeriod("frozen" /* PageState.FROZEN */, vital.startClocks.relative, vital.duration);
    }
    function addDurationVital(vital) {
        if (isValid(vital)) {
            lifeCycle.notify(12 /* LifeCycleEventType.RAW_RUM_EVENT_COLLECTED */, processVital(vital));
        }
    }
    function addOperationStepVital(name, stepType, options, failureReason) {
        if (!(0, browser_core_1.isExperimentalFeatureEnabled)(browser_core_1.ExperimentalFeature.FEATURE_OPERATION_VITAL)) {
            return;
        }
        const { operationKey, context, description } = options || {};
        const vital = {
            name,
            type: rawRumEvent_types_1.VitalType.OPERATION_STEP,
            operationKey,
            failureReason,
            stepType,
            startClocks: (0, browser_core_1.clocksNow)(),
            context: (0, browser_core_1.sanitize)(context),
            description,
        };
        lifeCycle.notify(12 /* LifeCycleEventType.RAW_RUM_EVENT_COLLECTED */, processVital(vital));
    }
    return {
        addOperationStepVital,
        addDurationVital,
        startDurationVital: (name, options = {}) => startDurationVital(customVitalsState, name, options),
        stopDurationVital: (nameOrRef, options = {}) => {
            stopDurationVital(addDurationVital, customVitalsState, nameOrRef, options);
        },
    };
}
function startDurationVital({ vitalsByName, vitalsByReference }, name, options = {}) {
    const vital = {
        name,
        startClocks: (0, browser_core_1.clocksNow)(),
        ...options,
    };
    // To avoid leaking implementation details of the vital, we return a reference to it.
    const reference = { __dd_vital_reference: true };
    vitalsByName.set(name, vital);
    // To avoid memory leaks caused by the creation of numerous references (e.g., from improper useEffect implementations), we use a WeakMap.
    vitalsByReference.set(reference, vital);
    return reference;
}
function stopDurationVital(stopCallback, { vitalsByName, vitalsByReference }, nameOrRef, options = {}) {
    const vitalStart = typeof nameOrRef === 'string' ? vitalsByName.get(nameOrRef) : vitalsByReference.get(nameOrRef);
    if (!vitalStart) {
        return;
    }
    stopCallback(buildDurationVital(vitalStart, vitalStart.startClocks, options, (0, browser_core_1.clocksNow)()));
    if (typeof nameOrRef === 'string') {
        vitalsByName.delete(nameOrRef);
    }
    else {
        vitalsByReference.delete(nameOrRef);
    }
}
function buildDurationVital(vitalStart, startClocks, stopOptions, stopClocks) {
    var _a;
    return {
        name: vitalStart.name,
        type: rawRumEvent_types_1.VitalType.DURATION,
        startClocks,
        duration: (0, browser_core_1.elapsed)(startClocks.timeStamp, stopClocks.timeStamp),
        context: (0, browser_core_1.combine)(vitalStart.context, stopOptions.context),
        description: (_a = stopOptions.description) !== null && _a !== void 0 ? _a : vitalStart.description,
        handlingStack: vitalStart.handlingStack,
    };
}
function processVital(vital) {
    const { startClocks, type, name, description, context, handlingStack } = vital;
    const vitalData = {
        id: (0, browser_core_1.generateUUID)(),
        type,
        name,
        description,
        ...(type === rawRumEvent_types_1.VitalType.DURATION
            ? { duration: (0, browser_core_1.toServerDuration)(vital.duration) }
            : {
                step_type: vital.stepType,
                operation_key: vital.operationKey,
                failure_reason: vital.failureReason,
            }),
    };
    return {
        rawRumEvent: {
            date: startClocks.timeStamp,
            vital: vitalData,
            type: rawRumEvent_types_1.RumEventType.VITAL,
            context,
        },
        startClocks,
        duration: type === rawRumEvent_types_1.VitalType.DURATION ? vital.duration : undefined,
        domainContext: handlingStack ? { handlingStack } : {},
    };
}
