// @datadog/browser-rum-core/cjs/domain/resource/trackManualResources.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.trackManualResources = trackManualResources;
const browser_core_1 = require("@datadog/browser-core");
const rawRumEvent_types_1 = require("../../rawRumEvent.types");
const resourceUtils_1 = require("./resourceUtils");
function trackManualResources(lifeCycle, resourceTracker) {
    function emitResource(id, startClocks, data, statusCode, endClocks) {
        const duration = endClocks ? (0, browser_core_1.elapsed)(startClocks.relative, endClocks.relative) : undefined;
        const rawRumEvent = {
            date: startClocks.timeStamp,
            type: rawRumEvent_types_1.RumEventType.RESOURCE,
            resource: {
                id,
                type: data.type || browser_core_1.ResourceType.OTHER,
                url: (0, resourceUtils_1.sanitizeIfLongDataUrl)(data.url),
                duration: duration !== undefined ? (0, browser_core_1.toServerDuration)(duration) : undefined,
                method: data.method,
                status_code: statusCode,
            },
            _dd: {},
            context: data.context,
        };
        lifeCycle.notify(12 /* LifeCycleEventType.RAW_RUM_EVENT_COLLECTED */, {
            rawRumEvent,
            startClocks,
            duration,
            domainContext: { isManual: true },
        });
    }
    function startManualResource(url, options = {}, startClocks = (0, browser_core_1.clocksNow)()) {
        var _a;
        const lookupKey = (_a = options.resourceKey) !== null && _a !== void 0 ? _a : url;
        resourceTracker.start(lookupKey, startClocks, {
            url,
            ...options,
        });
    }
    function stopManualResource(url, options = {}, stopClocks = (0, browser_core_1.clocksNow)()) {
        var _a;
        const lookupKey = (_a = options.resourceKey) !== null && _a !== void 0 ? _a : url;
        const stopped = resourceTracker.stop(lookupKey, stopClocks, {
            context: options.context,
            type: options.type,
        });
        if (!stopped) {
            return;
        }
        emitResource(stopped.id, stopped.startClocks, stopped, options.statusCode, stopClocks);
    }
    return {
        startResource: startManualResource,
        stopResource: stopManualResource,
    };
}
