// @datadog/browser-rum/esm/domain/record/internalApi.js
import { noop, timeStampNow } from '@datadog/browser-core';
import { getNodePrivacyLevel, NodePrivacyLevel } from '@datadog/browser-rum-core';
import { takeFullSnapshot as doTakeFullSnapshot } from './startFullSnapshots';
import { createRecordingScope } from './recordingScope';
import { createElementsScrollPositions } from './elementsScrollPositions';
import { serializeInTransaction, serializeNode } from './serialization';
/**
 * Take a full snapshot of the document, generating the same records that the browser SDK
 * would generate.
 *
 * This is an internal API function. Be sure to update Datadog-internal callers if you
 * change its signature or behavior.
 */
export function takeFullSnapshot({ configuration, } = {}) {
    const records = [];
    const emitRecord = (record) => {
        records.push(record);
    };
    doTakeFullSnapshot(timeStampNow(), 0 /* SerializationKind.INITIAL_FULL_SNAPSHOT */, emitRecord, noop, createTemporaryRecordingScope(configuration));
    return records;
}
/**
 * Take a snapshot of a DOM node, generating the serialized representation that the
 * browser SDK would generate.
 *
 * This is an internal API function. Be sure to update Datadog-internal callers if you
 * change its signature or behavior.
 */
export function takeNodeSnapshot(node, { configuration } = {}) {
    let serializedNode = null;
    serializeInTransaction(0 /* SerializationKind.INITIAL_FULL_SNAPSHOT */, noop, noop, createTemporaryRecordingScope(configuration), (transaction) => {
        const privacyLevel = getNodePrivacyLevel(node, transaction.scope.configuration.defaultPrivacyLevel);
        if (privacyLevel === NodePrivacyLevel.HIDDEN || privacyLevel === NodePrivacyLevel.IGNORE) {
            return;
        }
        serializedNode = serializeNode(node, privacyLevel, transaction);
    });
    return serializedNode;
}
function createTemporaryRecordingScope(configuration) {
    return createRecordingScope({
        defaultPrivacyLevel: NodePrivacyLevel.ALLOW,
        ...configuration,
    }, createElementsScrollPositions(), {
        addShadowRoot: noop,
        removeShadowRoot: noop,
        flush: noop,
        stop: noop,
    });
}
