// @datadog/browser-rum/esm/domain/record/startFullSnapshots.js
import { getViewportDimension } from '@datadog/browser-rum-core';
import { ExperimentalFeature, isExperimentalFeatureEnabled, timeStampNow } from '@datadog/browser-core';
import { RecordType } from '../../types';
import { serializeFullSnapshotAsChange, serializeFullSnapshot } from './serialization';
import { getVisualViewport } from './viewports';
export function startFullSnapshots(lifeCycle, emitRecord, emitStats, flushMutations, scope, serialize = defaultSerializeFullSnapshotCallback()) {
    takeFullSnapshot(timeStampNow(), 0 /* SerializationKind.INITIAL_FULL_SNAPSHOT */, emitRecord, emitStats, scope, serialize);
    const { unsubscribe } = lifeCycle.subscribe(2 /* LifeCycleEventType.VIEW_CREATED */, (view) => {
        flushMutations();
        takeFullSnapshot(view.startClocks.timeStamp, 1 /* SerializationKind.SUBSEQUENT_FULL_SNAPSHOT */, emitRecord, emitStats, scope, serialize);
    });
    return {
        stop: unsubscribe,
    };
}
export function takeFullSnapshot(timestamp, kind, emitRecord, emitStats, scope, serialize = defaultSerializeFullSnapshotCallback()) {
    const { width, height } = getViewportDimension();
    emitRecord({
        data: {
            height,
            href: window.location.href,
            width,
        },
        type: RecordType.Meta,
        timestamp,
    });
    emitRecord({
        data: {
            has_focus: document.hasFocus(),
        },
        type: RecordType.Focus,
        timestamp,
    });
    serialize(timestamp, kind, document, emitRecord, emitStats, scope);
    if (window.visualViewport) {
        emitRecord({
            data: getVisualViewport(window.visualViewport),
            type: RecordType.VisualViewport,
            timestamp,
        });
    }
}
function defaultSerializeFullSnapshotCallback() {
    return isExperimentalFeatureEnabled(ExperimentalFeature.USE_CHANGE_RECORDS)
        ? serializeFullSnapshotAsChange
        : serializeFullSnapshot;
}
